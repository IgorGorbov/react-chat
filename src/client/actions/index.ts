import {
  CLIENT_CONNECT,
  REGISTER_CLIENT,
  FETCH_ALL_USERS,
  FETCH_USER_CHATS,
  SEARCH_CHAT,
  CHANGE_USER_STATUS,
  CHANGE_CHAT,
  ADD_CHAT,
  NEW_CHAT,
  DELETE_CHAT,
  ADD_MESSAGE,
  NEW_MESSAGE,
  CHOOSE_USER,
  SELECT_MESSAGE,
  DELETE_MESSAGES,
  READ_MESSAGES,
  REFRESH_MESSAGES,
} from '../constants/actionTypes';

export const clientConnect = () => ({
  type: CLIENT_CONNECT,
});

export const changeUserStatus = (status: string) => ({
  type: CHANGE_USER_STATUS,
  payload: status,
});

export const changeChat = (chat: { id: number; companion: IUser }) => ({
  type: CHANGE_CHAT,
  payload: {
    id: chat.id,
    companion: chat.companion,
  },
});

export const fetchAllUsers = (users: IUser[]) => ({
  type: FETCH_ALL_USERS,
  payload: users,
});

export const fetchUserChats = (chats: IChat[]) => ({
  type: FETCH_USER_CHATS,
  payload: chats,
});

export const search = (text: string) => ({
  type: SEARCH_CHAT,
  payload: text,
});

export const addChat = (companion: IUser) => ({
  type: ADD_CHAT,
  payload: companion,
});

export const newChat = (currentUser: IUser, companion: IUser) => ({
  type: NEW_CHAT,
  payload: { currentUser, companion },
});

export const deleteChat = (chatId: number) => ({
  type: DELETE_CHAT,
  payload: chatId,
});

export const addMessage = (text: string) => ({
  type: ADD_MESSAGE,
  payload: text,
});

export const newMessage = (chatId: number, message: IMessage) => ({
  type: NEW_MESSAGE,
  payload: {
    chatId,
    message,
  },
});

export const registerClient = (name: string) => ({
  type: REGISTER_CLIENT,
  payload: name,
});

export const chooseUser = (user: IUser) => ({
  type: CHOOSE_USER,
  payload: user,
});

export const selectMessage = (messageId: number) => ({
  type: SELECT_MESSAGE,
  payload: messageId,
});

export const deleteMessages = (
  selectedMessages: number[],
  currentUserID: number,
  currentChatId: number,
) => ({
  type: DELETE_MESSAGES,
  payload: { selectedMessages, currentUserID, currentChatId },
});

export const refreshMessages = (messages: number[], currentChatId: number) => ({
  type: REFRESH_MESSAGES,
  payload: { messages, currentChatId },
});

export const readMessages = (currentChatId: number, currentUserID: number) => ({
  type: READ_MESSAGES,
  payload: { currentUserID, currentChatId },
});
