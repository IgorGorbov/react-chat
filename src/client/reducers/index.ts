import { combineReducers } from 'redux';
import session from './session';
import chats from './chats';
import filters from './filters';
import currentChat from './currentChat';
import users from './users';

export default combineReducers({
  session,
  users,
  chats,
  currentChat,
  filters,
});
