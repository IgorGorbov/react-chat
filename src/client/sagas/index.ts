import { SagaIterator } from 'redux-saga';
import { all, takeEvery, call, select } from 'redux-saga/effects';
import * as types from '../constants/actionTypes';
import {
  getCompanion,
  getChatCurrentId,
  getCurrentUser,
} from '../selectors/index';

const clientConnect = function*(socket: ISocket): SagaIterator {
  yield call(socket.getUsers, 'available', (err: any, users: IUser[]) => {
    if (err) {
      console.error('clientConnect saga => getUsers', err);
      return;
    }
    socket.dispatchAction({ type: types.FETCH_ALL_USERS, payload: users });
  });
};

const registerClient = function*(
  socket: ISocket,
  action: { payload: string },
): SagaIterator {
  yield call(socket.register, action.payload, (err: any, user: IUser) => {
    if (err) {
      console.error('registerClient saga => register', err);
      return;
    }
    socket.dispatchAction({ type: types.CHOOSE_USER, payload: user });
  });

  yield call(socket.getUserChats, (err: any, chats: IChat[]) => {
    if (err) {
      console.error('registerClient saga => getUserChats', err);
      return;
    }
    socket.dispatchAction({ type: types.FETCH_USER_CHATS, payload: chats });
  });

  yield call(socket.getUsers, 'all', (err: any, users: IUser[]) => {
    if (err) {
      console.error('registerClient saga => getUsers', err);
      return;
    }
    socket.dispatchAction({ type: types.FETCH_ALL_USERS, payload: users });
  });
};

const addChat = function*(
  socket: ISocket,
  action: { payload: IUser },
): SagaIterator {
  yield call(socket.addChat, action.payload, (err: any, chat: IChat) => {
    if (err) {
      console.error('addChat saga => addChat', err);
      return;
    }
    socket.dispatchAction({ type: types.NEW_CHAT, payload: chat });
  });
};

const deleteChat = function*(
  socket: ISocket,
  action: { payload: number },
): SagaIterator {
  yield call(socket.deleteChat, action.payload);
};

const addMessage = function*(
  socket: ISocket,
  action: { payload: string },
): SagaIterator {
  const text = action.payload;
  const companion = yield select(getCompanion);
  const chatId = yield select(getChatCurrentId);

  yield call(socket.addMessage, text, chatId, companion);
};

const deleteMessages = function*(
  socket: ISocket,
  action: { payload: number[] },
): SagaIterator {
  const currentChatId = yield select(getChatCurrentId);

  yield call(
    socket.deleteMessages,
    action.payload,
    (err: any, messages: IMessage[]) => {
      if (err) {
        console.error('deleteMessages saga => deleteMessages', err);
        return;
      }
      socket.dispatchAction({
        type: types.REFRESH_MESSAGES,
        payload: { currentChatId, messages },
      });
    },
  );
};

const changeUserStatus = function*(socket: ISocket): SagaIterator {
  const currentUser = yield select(getCurrentUser);
  yield call(socket.changeUserStatus, currentUser);
};

export default function* SocketSaga(socket: ISocket): SagaIterator {
  yield all([
    takeEvery(types.CLIENT_CONNECT, clientConnect, socket),
    takeEvery(types.REGISTER_CLIENT, registerClient, socket),
    takeEvery(types.CHANGE_USER_STATUS, changeUserStatus, socket),
    takeEvery(types.ADD_CHAT, addChat, socket),
    takeEvery(types.DELETE_CHAT, deleteChat, socket),
    takeEvery(types.ADD_MESSAGE, addMessage, socket),
    takeEvery(types.DELETE_MESSAGES, deleteMessages, socket),
  ]);
}
