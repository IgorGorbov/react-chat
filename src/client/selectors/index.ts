import { createSelector } from 'reselect';

const getSearchText = (state: IState) =>
  state.filters.get('search').toLowerCase();

const getChats = (state: IState) => state.chats;

const getCurrentId = (state: IState) => state.session.get('id');

export const getCurrentUser = (state: IState) => state.session.toJS();

export const getChatCurrentId = (state: IState) => state.currentChat.get('id');

export const getCurrentChat = (state: IState) => state.currentChat.toJS();

export const getCompanion = (state: IState) =>
  state.currentChat.get('companion');

export const getSelectedMessages = (state: IState) =>
  state.filters.get('selectedMessages');

export const getSearchEntities: any = createSelector(
  getSearchText,
  getChats,
  getCurrentId,
  (searchText, chats: any, currentID) => {
    const searchByName = (chat: any) => {
      const user = chat.users.find((u: any) => u.id !== currentID);
      return user.name
        .trim()
        .toLowerCase()
        .includes(searchText);
    };
    return chats.filter(searchByName).toJS();
  },
);

export const getChatById: any = createSelector(
  getChats,
  getChatCurrentId,
  (chats, id) => {
    return chats.find((c: any) => c.id === id);
  },
);

const getUsers = (state: any) => state.users;

export const getAvailableUsers: any = createSelector(
  getChats,
  getUsers,
  getCurrentId,
  (chats: any, users, currentUserId) => {
    const filterUsers = Array.from(
      new Set(
        Object.values(chats.toJS()).reduce(
          (acc: any, chat: any) =>
            acc.concat(chat.users.map((user: any) => user.id)),
          [currentUserId],
        ),
      ),
    );

    return users.filter((user: any) => !filterUsers.includes(user.id));
  },
);
