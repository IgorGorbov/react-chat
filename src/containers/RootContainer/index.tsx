import * as React from 'react';
import { connect } from 'react-redux';
import CurrentChatContainer from '../../containers/CurrentChatContainer';
import TypingMessageContainer from '../TypingMessageContainer';
import SectionContacts from '../../components/SectionContacts';
import ModalWindow from '../../components/ModalLogin';
import { GridLayout } from '../../styledComponents';
import {
  clientConnect,
  chooseUser,
  fetchAllUsers,
  registerClient,
} from '../../actions';

interface IRootContainerProps {
  users: IUser[];
  currentUser: IUser | {};
  currentChat: ICurrentChat;
  clientConnect: () => { type: string };
  registerClient: (name: string) => { type: string; payload: string };
  fetchAllUsers: (users: IUser[]) => { type: string; payload: IUser[] };
  chooseUser: (user: IUser) => { type: string; payload: IUser };
}

class RootContainer extends React.Component<IRootContainerProps> {
  public componentWillMount(): void {
    this.props.clientConnect();
  }
  public render(): JSX.Element {
    // tslint:disable: no-shadowed-variable
    const { users, currentUser, currentChat, registerClient } = this.props;
    const withChat: number = Object.keys(currentChat).length;
    if (!currentUser || !Object.keys(currentUser).length) {
      return (
        <ModalWindow
          modalOpen={true}
          registerClient={registerClient}
          users={users}
        />
      );
    }
    return (
      <React.Fragment>
        <GridLayout className={withChat ? 'withChat' : ''}>
          <SectionContacts />
          <CurrentChatContainer />
          <TypingMessageContainer />
        </GridLayout>
      </React.Fragment>
    );
  }
}

const mapStateToProps = (state: IState) => ({
  currentUser: state.session.toJS(),
  currentChat: state.currentChat.toJS(),
  users: state.users,
});

const mapDispatchToProps: {
  clientConnect: () => { type: string };
  registerClient: (name: string) => { type: string; payload: string };
  fetchAllUsers: (users: IUser[]) => { type: string; payload: IUser[] };
  chooseUser: (user: IUser) => { type: string; payload: IUser };
} = {
  clientConnect,
  registerClient,
  fetchAllUsers,
  chooseUser,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(RootContainer);
