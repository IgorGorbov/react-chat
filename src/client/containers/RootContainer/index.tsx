import * as React from 'react';
import { connect } from 'react-redux';
import CurrentChatContainer from '../CurrentChatContainer';
import TypingMessageContainer from '../TypingMessageContainer';
import SectionContacts from '../../components/SectionContacts';
import ModalWindow from '../../components/ModalLogin';
import { GridLayout } from '../../styledComponents';
import {
  clientConnect,
  chooseUser,
  fetchAllUsers,
  registerClient,
} from '../../actions';
import { getCurrentUser, getCurrentChat } from '../../selectors';

interface IRootContainerProps {
  users: IUser[];
  currentUser: IUser | {};
  currentChat: ICurrentChat;
  clientConnect: () => { type: string };
  registerClient: (name: string) => { type: string; payload: string };
  fetchAllUsers: (users: IUser[]) => { type: string; payload: IUser[] };
  chooseUser: (user: IUser) => { type: string; payload: IUser };
}

class RootContainer extends React.Component<IRootContainerProps> {
  public componentWillMount(): void {
    this.props.clientConnect();
  }
  public render(): JSX.Element {
    // tslint:disable: no-shadowed-variable
    const { users, currentUser, currentChat, registerClient } = this.props;
    const withChat: number = Object.keys(currentChat).length;
    if (!currentUser || !Object.keys(currentUser).length) {
      return (
        <ModalWindow
          modalOpen={true}
          registerClient={registerClient}
          users={users}
        />
      );
    }
    return (
      <React.Fragment>
        <GridLayout className={withChat ? 'withChat' : ''}>
          <SectionContacts />
          <CurrentChatContainer />
          <TypingMessageContainer />
        </GridLayout>
      </React.Fragment>
    );
  }
}

const mapStateToProps = (state: IState) => ({
  currentUser: getCurrentUser(state),
  currentChat: getCurrentChat(state),
  users: state.users,
});

const mapDispatchToProps: {
  clientConnect: () => { type: string };
  registerClient: (name: string) => { type: string; payload: string };
  fetchAllUsers: (users: IUser[]) => { type: string; payload: IUser[] };
  chooseUser: (user: IUser) => { type: string; payload: IUser };
} = {
  clientConnect,
  registerClient,
  fetchAllUsers,
  chooseUser,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(RootContainer);
