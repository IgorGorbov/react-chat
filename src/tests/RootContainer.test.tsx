import * as React from 'react';
import toJSON from 'enzyme-to-json';
import { mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

import RootContainer from '../client/containers/RootContainer';
import ModalWindow from '../client/components/ModalLogin';
import { GridLayout } from '../client/styledComponents';
import { Map } from 'immutable';

describe('testing RootContainer', () => {
  interface IStore {
    session: Map<string, string | number>;
    chats: Map<string, IChat>;
    currentChat: Map<string, string>;
    filters: Map<string, string | number[]>;
    users: IUser[];
  }
  const initState: IStore = {
    session: Map({}),
    currentChat: Map({}),
    users: [],
    chats: Map({}),
    filters: Map({ search: '', selectedMessages: [] }),
  };
  const initStateWithUser: IStore = {
    session: Map({
      id: 1,
      name: 'test',
      avatar: 'test',
      status: 'offline',
    }),
    currentChat: Map({}),
    users: [],
    chats: Map({}),
    filters: Map({ search: '', selectedMessages: [] }),
  };

  const mockStore = configureStore();

  it('should contain ModalWindow', () => {
    const store = mockStore(initState);
    const wrapper = mount(
      <Provider store={store}>
        <RootContainer />
      </Provider>,
    );
    expect(wrapper.find(ModalWindow)).toHaveLength(1);
  });

  it('should contain GridLayout', () => {
    const store = mockStore(initStateWithUser);
    const wrapper = mount(
      <Provider store={store}>
        <RootContainer />
      </Provider>,
    );
    expect(wrapper.find(GridLayout)).toHaveLength(1);
  });

  it('snapshot RootContainer', () => {
    const store = mockStore(initState);
    const wrapper = mount(
      <Provider store={store}>
        <RootContainer />
      </Provider>,
    );
    const tree = toJSON(wrapper);
    expect(tree).toMatchSnapshot();
  });

  it('snapshot RootContainer with user', () => {
    const store = mockStore(initStateWithUser);
    const wrapper = mount(
      <Provider store={store}>
        <RootContainer />
      </Provider>,
    );
    const tree = toJSON(wrapper);
    expect(tree).toMatchSnapshot();
  });
});
