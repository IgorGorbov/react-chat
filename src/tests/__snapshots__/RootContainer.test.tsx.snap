// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing RootContainer snapshot RootContainer 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(RootContainer)>
    <RootContainer
      chooseUser={[Function]}
      clientConnect={[Function]}
      currentChat={Object {}}
      currentUser={Object {}}
      fetchAllUsers={[Function]}
      registerClient={[Function]}
      users={Array []}
    >
      <ModalLogin
        modalOpen={true}
        registerClient={[Function]}
        users={Array []}
      >
        <Modal
          centered={true}
          closeOnDimmerClick={true}
          closeOnDocumentClick={false}
          dimmer={true}
          eventPool="Modal"
          open={true}
        >
          <Portal
            closeOnDocumentClick={false}
            closeOnEscape={true}
            eventPool="Modal"
            mountNode={
              <body
                class="dimmable dimmed"
              >
                <div
                  class="ui page modals dimmer transition visible active"
                  style="display: flex;"
                >
                  <div
                    class="ui modal transition visible active"
                  >
                    <div
                      class="header"
                    >
                      Choice a character
                    </div>
                    <div
                      class="content"
                    >
                      <div
                        class="ui vertically divided five column grid"
                      >
                        <div
                          class="row"
                        />
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="ui page modals dimmer transition visible active"
                  style="display: flex;"
                >
                  <div
                    class="ui modal transition visible active"
                  >
                    <div
                      class="header"
                    >
                      Choice a character
                    </div>
                    <div
                      class="content"
                    >
                      <div
                        class="ui vertically divided five column grid"
                      >
                        <div
                          class="row"
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </body>
            }
            onClose={[Function]}
            onMount={[Function]}
            onOpen={[Function]}
            onUnmount={[Function]}
            open={true}
            openOnTriggerClick={true}
          >
            <PortalInner
              key="inner"
              mountNode={
                <body
                  class="dimmable dimmed"
                >
                  <div
                    class="ui page modals dimmer transition visible active"
                    style="display: flex;"
                  >
                    <div
                      class="ui modal transition visible active"
                    >
                      <div
                        class="header"
                      >
                        Choice a character
                      </div>
                      <div
                        class="content"
                      >
                        <div
                          class="ui vertically divided five column grid"
                        >
                          <div
                            class="row"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="ui page modals dimmer transition visible active"
                    style="display: flex;"
                  >
                    <div
                      class="ui modal transition visible active"
                    >
                      <div
                        class="header"
                      >
                        Choice a character
                      </div>
                      <div
                        class="content"
                      >
                        <div
                          class="ui vertically divided five column grid"
                        >
                          <div
                            class="row"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </body>
              }
              onMount={[Function]}
              onUnmount={[Function]}
            >
              <Ref
                innerRef={[Function]}
              >
                <div
                  className="ui page modals dimmer transition visible active"
                >
                  <Ref
                    innerRef={[Function]}
                  >
                    <div
                      className="ui modal transition visible active"
                      style={
                        Object {
                          "marginTop": null,
                        }
                      }
                    >
                      <MountNode
                        className="dimmable dimmed"
                      />
                      <ModalHeader>
                        <div
                          className="header"
                        >
                          Choice a character
                        </div>
                      </ModalHeader>
                      <ModalContent>
                        <div
                          className="content"
                        >
                          <Grid
                            columns="five"
                            divided="vertically"
                          >
                            <div
                              className="ui vertically divided five column grid"
                            >
                              <GridRow>
                                <div
                                  className="row"
                                />
                              </GridRow>
                            </div>
                          </Grid>
                        </div>
                      </ModalContent>
                    </div>
                  </Ref>
                </div>
              </Ref>
            </PortalInner>
          </Portal>
        </Modal>
      </ModalLogin>
    </RootContainer>
  </Connect(RootContainer)>
</Provider>
`;

exports[`testing RootContainer snapshot RootContainer with user 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(RootContainer)>
    <RootContainer
      chooseUser={[Function]}
      clientConnect={[Function]}
      currentChat={Object {}}
      currentUser={
        Object {
          "avatar": "test",
          "id": 1,
          "name": "test",
          "status": "offline",
        }
      }
      fetchAllUsers={[Function]}
      registerClient={[Function]}
      users={Array []}
    >
      <styled.div
        className=""
      >
        <div
          className="sc-bdVaJa kOPfQt"
        >
          <SectionContacts>
            <styled.section>
              <section
                className="sc-bwzfXH kvRyKt"
              >
                <Connect(UserInfoContainers)>
                  <UserInfoContainers
                    changeUserStatus={[Function]}
                    user={
                      Object {
                        "avatar": "test",
                        "id": 1,
                        "name": "test",
                        "status": "offline",
                      }
                    }
                  >
                    <UserInfo
                      changeUserStatus={[Function]}
                      user={
                        Object {
                          "avatar": "test",
                          "id": 1,
                          "name": "test",
                          "status": "offline",
                        }
                      }
                    >
                      <styled.div>
                        <div
                          className="sc-bxivhb vHrVv"
                        >
                          <Styled(Image)
                            bordersize="4px"
                            circular={true}
                            cursor="true"
                            maxheight="80px"
                            maxwidth="80px"
                            onClick={[Function]}
                            size="tiny"
                            src="test"
                          >
                            <Image
                              as="img"
                              bordersize="4px"
                              circular={true}
                              className="sc-ifAKCX jpxJao"
                              cursor="true"
                              maxheight="80px"
                              maxwidth="80px"
                              onClick={[Function]}
                              size="tiny"
                              src="test"
                              ui={true}
                            >
                              <img
                                bordersize="4px"
                                className="ui tiny circular image sc-ifAKCX jpxJao"
                                cursor="true"
                                maxheight="80px"
                                maxwidth="80px"
                                onClick={[Function]}
                                src="test"
                              />
                            </Image>
                          </Styled(Image)>
                          <styled.span
                            textColor="white"
                            textFontFamily="Titillium Web"
                            textLineHeight="1"
                            textSize="18px"
                            textWeight="500"
                          >
                            <span
                              className="sc-EHOje epARVh"
                            >
                              test
                            </span>
                          </styled.span>
                        </div>
                      </styled.div>
                    </UserInfo>
                  </UserInfoContainers>
                </Connect(UserInfoContainers)>
                <Connect(SearchInput)>
                  <SearchInput
                    search={[Function]}
                  >
                    <Styled(Input)
                      icon="users"
                      iconPosition="left"
                      inputbgc="#485F78"
                      inputcolor="white"
                      inputwidth="100%"
                      onChange={[Function]}
                      placeholder="Search users ..."
                    >
                      <Input
                        className="sc-bZQynM bxBkOr"
                        icon="users"
                        iconPosition="left"
                        inputbgc="#485F78"
                        inputcolor="white"
                        inputwidth="100%"
                        onChange={[Function]}
                        placeholder="Search users ..."
                        type="text"
                      >
                        <div
                          className="ui left icon input sc-bZQynM bxBkOr"
                          inputbgc="#485F78"
                          inputcolor="white"
                          inputwidth="100%"
                        >
                          <input
                            onChange={[Function]}
                            placeholder="Search users ..."
                            type="text"
                          />
                          <Icon
                            as="i"
                            name="users"
                          >
                            <i
                              aria-hidden="true"
                              className="users icon"
                            />
                          </Icon>
                        </div>
                      </Input>
                    </Styled(Input)>
                  </SearchInput>
                </Connect(SearchInput)>
                <Connect(UserListContainer)>
                  <UserListContainer
                    addChat={[Function]}
                    availableUsers={Array []}
                    changeChat={[Function]}
                    chats={Object {}}
                    currentUser={
                      Object {
                        "avatar": "test",
                        "id": 1,
                        "name": "test",
                        "status": "offline",
                      }
                    }
                    readMessages={[Function]}
                  >
                    <ChatsList
                      changeChat={[Function]}
                      chats={Object {}}
                      currentUserId={1}
                      readMessages={[Function]}
                    >
                      <styled.div>
                        <div
                          className="sc-cSHVUG cliaFh"
                        />
                      </styled.div>
                    </ChatsList>
                    <ModalAddChat
                      addChat={[Function]}
                      currentUser={
                        Object {
                          "avatar": "test",
                          "id": 1,
                          "name": "test",
                          "status": "offline",
                        }
                      }
                      users={Array []}
                    >
                      <Modal
                        centered={true}
                        closeOnDimmerClick={true}
                        closeOnDocumentClick={false}
                        dimmer={true}
                        eventPool="Modal"
                        onClose={[Function]}
                        open={false}
                        trigger={
                          <Styled(Button)
                            color="black"
                            icon={true}
                            labelPosition="right"
                            onClick={[Function]}
                          >
                            Add User
                            <Icon
                              as="i"
                              name="add user"
                            />
                          </Styled(Button)>
                        }
                      >
                        <Portal
                          closeOnDocumentClick={false}
                          closeOnEscape={true}
                          eventPool="Modal"
                          mountNode={
                            <body
                              class="dimmable dimmed"
                            >
                              <div
                                class="ui page modals dimmer transition visible active"
                                style="display: flex;"
                              >
                                <div
                                  class="ui modal transition visible active"
                                >
                                  <div
                                    class="header"
                                  >
                                    Choice a character
                                  </div>
                                  <div
                                    class="content"
                                  >
                                    <div
                                      class="ui vertically divided five column grid"
                                    >
                                      <div
                                        class="row"
                                      />
                                    </div>
                                  </div>
                                </div>
                              </div>
                              <div
                                class="ui page modals dimmer transition visible active"
                                style="display: flex;"
                              >
                                <div
                                  class="ui modal transition visible active"
                                >
                                  <div
                                    class="header"
                                  >
                                    Choice a character
                                  </div>
                                  <div
                                    class="content"
                                  >
                                    <div
                                      class="ui vertically divided five column grid"
                                    >
                                      <div
                                        class="row"
                                      />
                                    </div>
                                  </div>
                                </div>
                              </div>
                            </body>
                          }
                          onClose={[Function]}
                          onMount={[Function]}
                          onOpen={[Function]}
                          onUnmount={[Function]}
                          open={false}
                          openOnTriggerClick={true}
                          trigger={
                            <Styled(Button)
                              color="black"
                              icon={true}
                              labelPosition="right"
                              onClick={[Function]}
                            >
                              Add User
                              <Icon
                                as="i"
                                name="add user"
                              />
                            </Styled(Button)>
                          }
                        >
                          <Ref
                            innerRef={[Function]}
                            key="trigger"
                          >
                            <Styled(Button)
                              color="black"
                              icon={true}
                              labelPosition="right"
                              onBlur={[Function]}
                              onClick={[Function]}
                              onFocus={[Function]}
                              onMouseEnter={[Function]}
                              onMouseLeave={[Function]}
                            >
                              <Button
                                as="button"
                                className="sc-htpNat hAnfKq"
                                color="black"
                                icon={true}
                                labelPosition="right"
                                onBlur={[Function]}
                                onClick={[Function]}
                                onFocus={[Function]}
                                onMouseEnter={[Function]}
                                onMouseLeave={[Function]}
                                role="button"
                              >
                                <button
                                  className="ui black icon right labeled button sc-htpNat hAnfKq"
                                  onBlur={[Function]}
                                  onClick={[Function]}
                                  onFocus={[Function]}
                                  onMouseEnter={[Function]}
                                  onMouseLeave={[Function]}
                                  role="button"
                                >
                                  Add User
                                  <Icon
                                    as="i"
                                    name="add user"
                                  >
                                    <i
                                      aria-hidden="true"
                                      className="add user icon"
                                    />
                                  </Icon>
                                </button>
                              </Button>
                            </Styled(Button)>
                          </Ref>
                        </Portal>
                      </Modal>
                    </ModalAddChat>
                  </UserListContainer>
                </Connect(UserListContainer)>
              </section>
            </styled.section>
          </SectionContacts>
          <Connect(ContactHeaderContainer)>
            <ContactHeaderContainer
              currentUser={
                Object {
                  "avatar": "test",
                  "id": 1,
                  "name": "test",
                  "status": "offline",
                }
              }
              deleteChat={[Function]}
              deleteMessages={[Function]}
              readMessages={[Function]}
              selectMessage={[Function]}
              selectedMessages={Array []}
            />
          </Connect(ContactHeaderContainer)>
          <Connect(TypingMessageContainer)>
            <TypingMessageContainer
              addMessage={[Function]}
              currentUser={
                Object {
                  "avatar": "test",
                  "id": 1,
                  "name": "test",
                  "status": "offline",
                }
              }
            />
          </Connect(TypingMessageContainer)>
        </div>
      </styled.div>
    </RootContainer>
  </Connect(RootContainer)>
</Provider>
`;
